File: windows/AirMateWindows/MontFont.cs
using System.Drawing.Text;
using System.Runtime.InteropServices;

namespace AirMate;

/// <summary>Mont, carried inside the executable.</summary>
/// <remarks>
/// Installed fonts cannot be relied on and a private collection costs nothing, so all three
/// AirMate clients ship the same files and none is typeset by whatever happens to be on the
/// machine. The pinned bytes must outlive the collection that was handed their address, so they
/// are held for the life of the process rather than left to be collected — let them move and the
/// font vanishes mid-run.
/// </remarks>
internal static class MontFont
{
    private static readonly PrivateFontCollection Collection = new();
    private static readonly List<GCHandle> Pinned = [];

    static MontFont()
    {
        foreach (var name in new[] { "mont_thin.ttf", "mont_black.ttf" })
        {
            try
            {
                using var stream = typeof(MontFont).Assembly.GetManifestResourceStream(name);
                if (stream is null) continue;
                var bytes = new byte[stream.Length];
                stream.ReadExactly(bytes);
                var handle = GCHandle.Alloc(bytes, GCHandleType.Pinned);
                Pinned.Add(handle);
                Collection.AddMemoryFont(handle.AddrOfPinnedObject(), bytes.Length);
            }
            catch
            {
                // A missing weight falls back below rather than taking the window down with it.
            }
        }
    }

    public static Font Thin(float size) => Make("Mont Thin", size);
    public static Font Black(float size) => Make("Mont Black", size);

    private static Font Make(string name, float size)
    {
        var family = Collection.Families.FirstOrDefault(f => f.Name == name);
        if (family is not null)
        {
            try { return new Font(family, size, FontStyle.Regular, GraphicsUnit.Point); }
            catch { }
        }
        var fallback = (SystemFonts.MessageBoxFont ?? SystemFonts.DefaultFont).FontFamily;
        return new Font(fallback, size, name.EndsWith("Black") ? FontStyle.Bold : FontStyle.Regular);
    }
}

/// <summary>White carries all the hierarchy, through opacity alone.</summary>
internal static class Mont
{
    public static readonly Color Ground = Color.Black;
    public static readonly Color Active = Color.FromArgb(255, 255, 255, 255);
    public static readonly Color Primary = Color.FromArgb(235, 255, 255, 255);
    public static readonly Color Detail = Color.FromArgb(158, 255, 255, 255);
    public static readonly Color Dim = Color.FromArgb(148, 255, 255, 255);
    public static readonly Color Disabled = Color.FromArgb(89, 255, 255, 255);
    public static readonly Color Track = Color.FromArgb(23, 255, 255, 255);

    public static readonly Color Mustard = Color.FromArgb(0xD8, 0xA6, 0x28);
    public static readonly Color Live = Color.FromArgb(0x2E, 0x9E, 0x5B);
    public static readonly Color Danger = Color.FromArgb(0xC0, 0x39, 0x2B);

    /// <summary>Text hangs off a generous left margin and nothing needs the right one.</summary>
    public const int PadLeft = 22;
    public const int PadRight = 14;
}

File: mac/Sources/AirMateMac/MontFont.swift
import AppKit
import CoreText
import SwiftUI

/// Mont, loaded from the bundle. The same typeface the Android client uses, so the two halves of
/// AirMate do not look like they were made by different people.
///
/// Only Thin and Black ship here. Thin does one job — the `air` of the wordmark — and Black is
/// everything else: under Mont it is the default weight rather than an emphasis weight, which is
/// what lets a plain word act as a button without a box around it.
enum MontFont {
    static let thin = "Mont-Thin"
    static let black = "Mont-Black"

    static func register() {
        for name in ["mont_thin", "mont_black"] {
            guard let url = Bundle.main.url(forResource: name, withExtension: "ttf") else {
                Diagnostics.shared.displayLog.error("Mont \(name).ttf missing from the bundle")
                continue
            }
            CTFontManagerRegisterFontsForURL(url as CFURL, .process, nil)
        }
    }
}

extension Font {
    static func montBlack(_ size: CGFloat) -> Font { .custom(MontFont.black, size: size) }
    static func montThin(_ size: CGFloat) -> Font { .custom(MontFont.thin, size: size) }
}

/// White carries all the hierarchy, through opacity alone.
enum MontWhite {
    static let active = 1.0
    static let primary = 0.92
    static let detail = 0.62
    static let dim = 0.58
    static let disabled = 0.35
    static let track = 0.09
}

enum MontColor {
    /// The Mont surface: black, with whatever is behind it faintly present through the last
    /// eight percent. One value, shared with the Android client and with MiniMate.
    static let surface = Color.black.opacity(0.92)
    static let mustard = Color(red: 0xD8 / 255, green: 0xA6 / 255, blue: 0x28 / 255)
    static let live = Color(red: 0x2E / 255, green: 0x9E / 255, blue: 0x5B / 255)
    static let danger = Color(red: 0xC0 / 255, green: 0x39 / 255, blue: 0x2B / 255)
}
